protocols/bftcosi/bftcosi.go

```diff
@@ -0,0 +1,615 @@
+		bft.tpcMut.Lock()
+		bft.tempPrepareCommit = append(bft.tempPrepareCommit, comm.Commitment)
+		if len(bft.tempPrepareCommit) < len(bft.Children()) {
+			bft.tpcMut.Unlock()
+			return nil
+		}
+		commitment = bft.prepare.Commit(bft.tempPrepareCommit)
+		bft.tpcMut.Unlock()
+		if bft.IsRoot() {
+			return bft.startChallengePrepare()
+		}
+
+		dbg.Lvl3(bft.Name(), "BFTCoSi handle Commit PREPARE")
+	case RoundCommit:
+		bft.tccMut.Lock()
+		bft.tempCommitCommit = append(bft.tempCommitCommit, comm.Commitment)
+		if len(bft.tempCommitCommit) < len(bft.Children()) {
+			bft.tccMut.Unlock()
+			return nil
+		}
+		commitment = bft.commit.Commit(bft.tempCommitCommit)
+		bft.tccMut.Unlock()
+		if bft.IsRoot() {
+			// do nothing:
+			// stop the processing of the round, wait the end of
+			// the "prepare" round: calls startChallengeCommit
+			return nil
+		}
+
+		dbg.Lvl3(bft.Name(), "BFTCoSi handle Commit COMMIT")
+	}
+	// set same RoundType as for the received commitment:
+	typedCommitment.TYPE = comm.TYPE
+	typedCommitment.Commitment = commitment
+	err := bft.SendTo(bft.Parent(), typedCommitment)
+	return err
+}
+
+// startPrepareChallenge create the challenge and send its down the tree
+func (bft *ProtocolBFTCoSi) startChallengePrepare() error {
+	// create challenge from message's hash:
+	prep := &msgPrepare{Msg: bft.Msg}
+	h := prep.Hash()
+	ch, err := bft.prepare.CreateChallenge(h)
+	if err != nil {
+		return err
+	}
+
+	bftChal := &ChallengePrepare{
+		TYPE:      RoundPrepare,
+		Challenge: ch,
+		Msg:       bft.Msg,
+	}
+
+	go bft.verificationFun(bft.Msg, bft.verifyChan)
+
+	dbg.Lvl3(bft.Name(), "BFTCoSi Start Challenge PREPARE")
+	// send to children
+	for _, tn := range bft.Children() {
+		err = bft.SendTo(tn, bftChal)
+	}
+	return err
+}
+
+// startCommitChallenge waits the end of the "prepare" round.
+// Then it creates the challenge and sends it along with the
+// "prepare" signature down the tree.
+func (bft *ProtocolBFTCoSi) startChallengeCommit() error {
+	c, err := bft.commit.CreateChallenge(bft.Msg)
+	if err != nil {
+		return err
+	}
+
+	// send challenge + signature
+	cc := &ChallengeCommit{
+		TYPE:      RoundCommit,
+		Challenge: c,
+		Signature: bft.prepare.Signature(),
+	}
+	dbg.Lvl3("BFTCoSi Start Challenge COMMIT")
+	for _, tn := range bft.Children() {
+		err = bft.SendTo(tn, cc)
+	}
+	return err
+}
+
+// handlePrepareChallenge receive the challenge messages for the "prepare"
+// round.
+func (bft *ProtocolBFTCoSi) handleChallengePrepare(ch *ChallengePrepare) error {
+	bft.Msg = ch.Msg
+	// start the verification of the message
+	go bft.verificationFun(bft.Msg, bft.verifyChan)
+	// acknowledge the challenge and send its down
+	chal := bft.prepare.Challenge(ch.Challenge)
+	ch.Challenge = chal
+
+	dbg.Lvl3(bft.Name(), "BFTCoSi handle Challenge PREPARE")
+	// go to response if leaf
+	if bft.IsLeaf() {
+		return bft.startResponsePrepare()
+	}
+	var err error
+	for _, tn := range bft.Children() {
+		err = bft.SendTo(tn, ch)
```

Comment:
For me this could even be the default behaviour of the function.

Another question is whether the `SendTo` should go into a go-routine. Useful if you have lots of children. Specially in the set-up phase, where a connection can take quite some time until it is set up (n * 100ms)...